exercises/src/main/scala/fpinscala/laziness/Stream.scala

```diff
@@ -17,20 +27,95 @@ trait Stream[+A] {
     case Empty => None
     case Cons(h, t) => if (f(h())) Some(h()) else t().find(f)
   }
-  def take(n: Int): Stream[A] = ???
 
-  def drop(n: Int): Stream[A] = ???
+  //5.2
+  def take(n: Int): Stream[A] = this match {
+    case _ if n == 0 => empty
+    case Cons(h, _) if n == 1 => cons(h(), empty)
+    case Cons(h, t) => cons(h(), t().take(n - 1))
+  }
+
+  //5.2
+  def drop(n: Int): Stream[A] = this match {
+    case Cons(_, t) if n >= 1 => t().drop(n - 1)
+    case _ => this
+  }
+
+  //5.3
+  def takeWhile(p: A => Boolean): Stream[A] = this match {
+    case Cons(h, t) if p(h()) => cons(h(), t().takeWhile(p))
+    case _ => empty
+  }
 
-  def takeWhile(p: A => Boolean): Stream[A] = ???
+  //5.4
+  def forAll(p: A => Boolean): Boolean = foldRight(true)((a, b) => p(a) && b )
 
-  def forAll(p: A => Boolean): Boolean = ???
+  //5.5 takeWhile with foldRight
+  def takeWhile2(p: A => Boolean): Stream[A] = foldRight(Empty: Stream[A])((a, b) => if (p(a)) cons(a, b) else Empty)
 
-  def headOption: Option[A] = ???
+  //5.6 headOption with foldRight
+  def headOption: Option[A] = foldRight(None: Option[A])((a, _) => Some(a))
 
   // 5.7 map, filter, append, flatmap using foldRight. Part of the exercise is
   // writing your own function signatures.
+  def map[B](f: A => B): Stream[B] = foldRight(Empty: Stream[B])((a, b) => cons(f(a), b))
```

Comment:
nit: The book author added `def empty[A]: Stream[A] = Empty` in this file to make it easier to cast an empty stream. You can use `empty[B]` instead of `Empty: Stream[B]`.